File: pohlig_hellman.m
function x = pohlig_hellman(p, a, b)
    factors = factor(p-1);
    qlist = unique(factors);
    counterlist = histcounts(round(double(factors)));
    indices = find(counterlist == 0);
    counterlist(indices) = [];
    counterlist = sym(counterlist);
    masterx = [];
    for index=1:length(qlist)
     q = qlist(index);
     xlist = [];
     prevb = powermod(b, (p-1)/q, p);
     currx = prevb;
     for k=0:q-1
         if (powermod(a, k * ((p-1)/q), p) == currx)
             xlist = [xlist, k];
             break
         end
     end
     % Found x0 and will now continue on to find the rest of the current q’s
     % betas and xs
     prevb = b;
     counter = sym('1');
     while (counter < counterlist(index))
      
         prevb = mod(prevb*powermod(invmodn(powermod(a,xlist(end),p), p), q^(counter-1), p), p);
      
         currx = powermod(prevb, (p-1)/(q^(counter + 1)), p);
         for k=0:q-1
             if (powermod(a, k * ((p-1)/q), p) == currx)
                 xlist = [xlist, k];
                 break
             end
         end
         counter = counter + 1;
     end
     overallx = 0;
     % overallx is the x found for the current modulus
     for j=sym('1'):counterlist(index)
      
         overallx = mod(overallx + (q^(j-1))*xlist(j), q^counterlist(index));
     end
     % masterx holds all of the x for all respective modulus
     masterx = [masterx overallx];
    end
    xforcrt = [];
    for w=1:length(masterx)
     xforcrt = [xforcrt; [masterx(w), qlist(w)^counterlist(w)]];
    end
    % xforcrt holds all x and q^r pairings where r is the number of q that
    % factor into p-1

    x = crt(xforcrt(:, 1).', xforcrt(:, 2).');

end

File: Python/pohlig_hellman.py
import helper_functions
import sympy

def pohlig_hellman(p, a, b):
    factorMap = sympy.factorint(p-1)
    qlist = list(factorMap.keys())
    counterlist = list(factorMap.values())
    masterx = []
    for index in range(len(qlist)):
        q = qlist[index]
        xlist = []
        prevb = helper_functions.powermod(b, (p-1)//q, p)
        currx = prevb
        for k in range(q):
            if helper_functions.powermod(a, k * ((p-1)//q), p) == currx:
                xlist.append(k)
                break
        prevb = b
        counter = 1
        while counter < counterlist[index]:
            prevb = (prevb * helper_functions.powermod(helper_functions.invmodn(helper_functions.powermod(a, xlist[-1], p), p), (q**(counter - 1)), p) % p)
            currx = helper_functions.powermod(prevb, (p - 1) // (q**(counter + 1)), p)
            for k in range(q):
                if helper_functions.powermod(a, k * ((p-1)//q), p) == currx:
                    xlist.append(k)
                    break
            counter = counter + 1
        overallx = 0
        for j in range(1, counterlist[index]+1):
            overallx = (overallx + (q**(j-1))*xlist[j-1]) % (q**counterlist[index])
        masterx.append(overallx)
    xforcrt = []
    for w in range(len(masterx)):
        xforcrt.append((masterx[w], (qlist[w]**counterlist[w])))
    firstlist = []
    secondlist = []
    for tup in xforcrt:
        firstlist.append(tup[0])
        secondlist.append(tup[1])
    x = helper_functions.crt(firstlist, secondlist)
    return x
 
